tensorzero-core/src/endpoints/datasets.rs

```diff
@@ -1468,11 +1470,16 @@ where
     }
 }
 
-// SyntheticChatInferenceDatapoints are created by `update_datapoint_handler` and represent datapoints not derived from inferences. In storage their `source_inference_id` is null.
+// UpdateChatInferenceDatapointsReques is used in `update_datapoint_handler` to update an existing ChatInferenceDatapoint
+// in place. It's not a partial update - the entire request is written to the database.
+//
+// TODO(shuyangli): this is currently being manually kept in sync with `ChatInferenceDatapoint`. Can we fix this?
```

Comment:
we can do a two-way type assertion like we do [here](https://github.com/tensorzero/tensorzero/blob/d7a5dd59bcfdcf9266d8eed42a9b65f0e78161e8/clients/rust/src/client_inference_params.rs#L139). I think this should be a separate change and maybe subsumed by our larger API refactor but we do have a tool for this and I think it's a good point.